### PR TITLE
249r3

### PR DESCRIPTION
**Pull request by maxkazmsft**
Closes https://github.com/microsoft/seismic-deeplearning/issues/144 https://github.com/microsoft/seismic-deeplearning/issues/249

### CODE REVIEW

File: contrib/experiments/interpretation/dutchf3_patch/distributed/train.py [hunk 1]
```diff
@@ -155,12 +155,7 @@ def run(*options, cfg=None, local_rank=0, debug=False):
 
     logger.info(f"Validation examples {len(val_set)}")
     n_classes = train_set.n_classes
-
-    if debug:
-        logger.info("Running in debug mode..")
-        train_set = data.Subset(train_set, list(range(4)))
-        val_set = data.Subset(val_set, list(range(4)))
-
```
[the remark this answers is not in the document]
Comment (maxkazmsft) [reply contrib/experiments/interpretation/dutchf3_patch/distributed/train.py:163]: yeah will actually revert this change and possibly do a separate PR

File: experiments/interpretation/dutchf3_patch/local/train.py [hunk 2]
```diff
@@ -60,18 +60,23 @@ def run(*options, cfg=None, debug=False):
...
         cfg (str, optional): Location of config file to load. Defaults to None.
+        override (str, optional): Location of config file which loads updates to config file
+                                  loaded by the cfg parameter above, e.g. say you want to train
+                                  model on a reduced or test dataset
         debug (bool): Places scripts in debug/test mode and only executes a few iterations
     """
     # Configuration:
     update_config(config, options=options, config_file=cfg)
+    update_config(config, options=[], config_file=override)
+
```
[the remark this answers is not in the document]
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/train.py:71]: discussed - removing

File: scripts/gen_checkerboard.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    :param n_inlines: dim x
+    :param n_crosslines: dim y
+    :param n_depth: dim z
+    :return: numpy array
+    """
+    # inline by crossline by depth
+    zero_patch = np.ones((box_size, box_size)) * WHITE
+    one_patch = np.ones((box_size, box_size)) * BLACK
+
+    # multiplying by one flips the sign
```
Comment (yalaudah) [inline scripts/gen_checkerboard.py:39]:
```suggestion
    # multiplying by negative one flips the sign
```

File: scripts/gen_checkerboard.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    if not os.path.isdir(path):
+        os.mkdir(path)
+
+
+def main(args):
+    """
+
+    Generates checkerboard dataset based on Dutch F3 in Alaudah format.
+
+    TODO: please see the TODO statements in the code
```
Comment (yalaudah) [inline scripts/gen_checkerboard.py:87]: Infinite loop here :)

File: scripts/gen_checkerboard.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    test1_seismic = np.load(os.path.join(args.dataroot, "test_once", "test1_seismic.npy"))
+    test1_labels = np.load(os.path.join(args.dataroot, "test_once", "test1_labels.npy"))
+    test2_seismic = np.load(os.path.join(args.dataroot, "test_once", "test2_seismic.npy"))
+    test2_labels = np.load(os.path.join(args.dataroot, "test_once", "test2_labels.npy"))
+
+    assert train_seismic.shape == train_labels.shape
+    assert train_seismic.min() == WHITE
+    assert train_seismic.max() == BLACK
+    assert train_labels.min() == 0
+    assert train_labels.max() == 5
```
Comment (yalaudah) [inline scripts/gen_checkerboard.py:104]: Are 0 and 5 here related to the number of classes in the Dutchf3 datsset?

Comment (maxkazmsft) [reply scripts/gen_checkerboard.py:104]: Yes, added comment to address this

File: scripts/gen_checkerboard.py (hunk 4 above)
Comment (maxkazmsft) [reply scripts/gen_checkerboard.py:87]: LoL agreed - removed and updated description.

File: scripts/gen_checkerboard.py (hunk 3 above)
Comment (maxkazmsft) [reply scripts/gen_checkerboard.py:39]: yup updated

File: contrib/experiments/interpretation/dutchf3_patch/distributed/train.py [hunk 6]
```diff
@@ -156,13 +156,15 @@ def run(*options, cfg=None, local_rank=0, debug=False):
...
+        #train_set = data.Subset(train_set, range(config.TRAIN.BATCH_SIZE_PER_GPU*2))
+    
     logger.info(f"Training examples {len(train_set)}")
-    logger.info(f"Validation examples {len(val_set)}")
+    logger.info(f"Validation examples {len(val_set)}")    
+
+    train_sampler = torch.utils.data.distributed.DistributedSampler(train_set, num_replicas=world_size, rank=local_rank)
+
 
     train_sampler = torch.utils.data.distributed.DistributedSampler(train_set, num_replicas=world_size, rank=local_rank)
```
Comment (yalaudah) [inline contrib/experiments/interpretation/dutchf3_patch/distributed/train.py:169]: duplicate?

File: contrib/experiments/interpretation/dutchf3_patch/distributed/train.py [hunk 7]
```diff
@@ -197,8 +199,14 @@ def run(*options, cfg=None, local_rank=0, debug=False):
 
     model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[device], find_unused_parameters=True)
 
-    snapshot_duration = epochs_per_cycle * len(train_loader) if not debug else 2 * len(train_loader)
+    snapshot_duration = epochs_per_cycle * len(train_loader) if not debug else 2*len(train_loader)
+
+    snapshot_duration = scheduler_step * len(train_loader)
+    if debug:
+        snapshot_duration = 2
```
Comment (yalaudah) [inline contrib/experiments/interpretation/dutchf3_patch/distributed/train.py:206]: duplicate? I think the first one is the latest version

File: experiments/interpretation/dutchf3_patch/local/test.py [hunk 8]
```diff
@@ -414,17 +425,35 @@ def test(*options, cfg=None, debug=False):
...
-    logger.info(f'Mean IoU: {score["Mean IoU: "]:0.3f}')
+    logger.info(f'Pixel Acc: {score["Pixel Acc: "]:.4f}')
+
+    if debug:
+        for cdx in range(n_classes):
+            logger.info(f'  Class_{cdx}_accuracy {score["Class Accuracy: "][cdx]:.3f}')
+    else:
+        for cdx, class_name in enumerate(_CLASS_NAMES):
+            logger.info(f'     {class_name}_accuracy {score["Class Accuracy: "][cdx]:.4f}')
+
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/test.py:436]: this if debug statement and the one before are almost identical.. hard to see here without the whole file but is it a duplicate? can we have them all under the same if statement if they're not?

File: contrib/experiments/interpretation/dutchf3_patch/distributed/train.py (hunk 7 above)
Comment (maxkazmsft) [reply contrib/experiments/interpretation/dutchf3_patch/distributed/train.py:206]: bad merge - resolved.

File: experiments/interpretation/dutchf3_patch/local/test.py (hunk 8 above)
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/test.py:436]: one is per split and the other is for final results, agree that they look similar - without refactoring test.py can't do much about it... flag refactoring test.py as a separate github issue?

File: contrib/experiments/interpretation/dutchf3_patch/distributed/train.py (hunk 6 above)
Comment (maxkazmsft) [reply contrib/experiments/interpretation/dutchf3_patch/distributed/train.py:169]: yes bad merge - good catch!

**Pull request merged**